### ISSUE TITLE
Load weights from TensorFlow checkpoint to Keras model

### ISSUE THREAD

**Issue by AliOsm** [tensorflow/tensorflow#24624]
I have trained a TensorFlow with Keras model and using `keras.callbacks.ModelCheckpoint` I've saved the weights as follows:

    cp_callback = keras.callbacks.ModelCheckpoint(checkpoint_path,
                                              save_weights_only=True,
                                              verbose=1)
    model.fit(X_train, Y_train, callbacks=[cp_callback], epochs=50, batch_size=256)

But while trying to load the saved weights there is nothing changed to my model, after building the model architecture and compile it I load weights as follows:

    model.load_weights('./checkpoints/cp.ckpt')

But nothing happens, the testing accuracy is as random guessing, while my real testing accuracy is 80.49%

The model consists of keras dence layers with l2 kernel_regularizer and glorot kernel_initializer, also I'm using TensorFlow version 1.12.0, any ideas?

**Comment by kasperfred** [comment]
Do you use `tf.keras` or `keras` standalone?
 
Are you sure your `checkpoint_path` is correct?

Could you run the following script? 
```Python
import tensorflow as tf
print(tf.__version__)

mnist = tf.keras.datasets.mnist
(x_train, y_train), (x_test, y_test) = mnist.load_data()


def build_model():
    model = tf.keras.models.Sequential()

    model.add(tf.keras.layers.Flatten(input_shape=(28,28,)))
    model.add(tf.keras.layers.Dense(32, activation="relu"))
    model.add(tf.keras.layers.Dense(10, activation="sigmoid"))
    
    return model

model = build_model()

model.compile(optimizer=tf.train.AdamOptimizer(), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

print(model.summary())

cp_callback = tf.keras.callbacks.ModelCheckpoint("./weights.{epoch:02d}.hdf5",
                                          save_weights_only=True,
                                          verbose=1)
model.fit(x_train, y_train, callbacks=[cp_callback], epochs=3, batch_size=256)


# rebuild model as model2
# simulating starting a new script
model2 = build_model()
model2.compile(optimizer=tf.train.AdamOptimizer(), 
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

# check that the new model does indeed have random weights
print (model2.evaluate(x=x_test, y=y_test))

# load 3rd epoch weights
model2.load_weights('./weights.03.hdf5')

# these should be equal
print (model2.evaluate(x=x_test, y=y_test))
print (model.evaluate(x=x_test, y=y_test))
```

**Comment by AliOsm** [comment]
I'm working with `tf.keras` and the checkpoint_path is correct. Also, the output of the evaluation in your code is correct:

```
10000/10000 [==============================] - 0s 43us/step
[9.01510094833374, 0.1127]
10000/10000 [==============================] - 0s 39us/step
[1.2179515656471251, 0.674]
10000/10000 [==============================] - 0s 41us/step
[1.2179515656471251, 0.674]
```

I think I figure the problem, after I load the weights I test it with the same testing data but shuffled (not in the same order) is this effects the accuracy? because if I evaluate using the same order it gives me the same results of the original model.

Another thing is when I load the weights I can't start training from the point I've stopped in, any idea?

**Comment by kasperfred** [comment]
Since the accuracy metric is a reduce operation, the order of the data should not matter. 

It's important, however, that when you're shuffling, that you're preserving the same order between the features and labels.

When you compile the model, you lose the optimizer state meaning it will require a few iterations to dial the learning rate back in. This may be why you can't resume training from the point you've stopped.

While I'm very new here, I think there's an official policy that Github issues are only for Tensorflow bugs, and since this doesn't appear to be a bug in Tensorflow, you may have a better experience asking subsequent questions on Stackoverflow using the tensorflow tag.

**Comment by AliOsm** [comment]
Actually I've asked the same question on Stackoverflow here: https://stackoverflow.com/questions/53971139/load-weights-from-tensorflow-checkpoint-to-keras-model but there is no answer, the shuffling should be ok, but still I don't know why the accuracy decreases after loading the model.

**Comment by kasperfred** [comment]
Could you send a working code snippet that demonstrates this behavior?

**Comment by AliOsm** [comment]
This is the code, by now the code saves and restore the checkpoint weights correctly. But as I said before, the testing data should be in the same order and it is not working from the same point that ends in with the fitting process.

```
import os
import math
import pickle
import numpy as np
import tensorflow as tf

from tensorflow import keras
from tensorflow.keras.callbacks import Callback

def create_dense_layer(neurons):
    return keras.layers.Dense(neurons, activation='relu',
                                        kernel_initializer=keras.initializers.glorot_normal(seed=961),
                                        kernel_regularizer=keras.regularizers.l2(0.0001))

def create_model():
    model = tf.keras.models.Sequential([
        keras.layers.Dense(200, activation=tf.nn.relu, input_shape=(200,)),
        create_dense_layer(500),
        create_dense_layer(500),
        create_dense_layer(450),
        create_dense_layer(400),
        create_dense_layer(400),
        create_dense_layer(350),
        create_dense_layer(300),
        create_dense_layer(300),
        create_dense_layer(250),
        create_dense_layer(200),
        create_dense_layer(200),
        create_dense_layer(150),
        create_dense_layer(100),
        create_dense_layer(100),
        create_dense_layer(50),
        create_dense_layer(25),
        keras.layers.Dense(CLSS_NUM, activation=tf.nn.softmax)
    ])
    
    model.compile(loss=tf.keras.losses.categorical_crossentropy, optimizer=tf.train.AdagradOptimizer(learning_rate=0.0075), metrics=['accuracy'])
    
    return model

model = create_model()

class TestCallback(Callback):
    def __init__(self, test_data):
        self.test_data = test_data

    def on_epoch_end(self, epoch, logs={}):
        x, y = self.test_data
        loss, acc = self.model.evaluate(x, y, verbose=0, batch_size=256)
        print('\nTesting loss: {}, acc: {}'.format(loss, acc))

checkpoint_path = 'checkpoints/cp-{epoch:04d}.ckpt'
cp_callback = tf.keras.callbacks.ModelCheckpoint(checkpoint_path, verbose=0)

model.fit(X_train, Y_train, epochs=10, batch_size=256, callbacks=[cp_callback, TestCallback((X_test, Y_test))])

loss, acc = model.evaluate(X_test, Y_test, batch_size=256)
print(loss, acc)
```

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by ymodak** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by LeninGF** [comment]
> Actually I've asked the same question on Stackoverflow here: https://stackoverflow.com/questions/53971139/load-weights-from-tensorflow-checkpoint-to-keras-model but there is no answer, the shuffling should be ok, but still I don't know why the accuracy decreases after loading the model.

I am facing this behavior but I am thinking that when the callback saves the model or model weights to the h5 file it uses the train acc value to save, which could be logically bigger than the validation accuracy, if I am right comment

**Comment by AliOsm** [comment]
Hello @LeninGF, the problem I faced previously was about the construction of the vocabulary. I was using Python set without sorting it, so the vocabulary order within the training was different from the vocabulary order within the testing.
